refactor: unify instruction & opcode types

enitrat opened the pull request.

obatirou commented on crates/common/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+num-traits.workspace = true
+serde = { version = "1.0", features = ["derive"] }
+serde_json.workspace = true
+thiserror = "2.0"
```
```suggestion
thiserror.workspace = true
```

zmalatrax commented on crates/prover/src/components/multiple_constraints.rs:
```diff
@@ -34,7 +35,7 @@ impl<const N: usize> Claim<N> {
         let mut trace = unsafe { ComponentTrace::<N>::uninitialized(self.log_size) };
-        let M31_0 = PackedM31::broadcast(M31::from(0));
+        let M31_0 = PackedM31::broadcast(Zero::zero());
         let M31_1 = PackedM31::broadcast(M31::from(1));
```
`use num_traits::One;`
`M31::one()`

Or `One::one()` if you keep the `Zero::zero()` to be consistent

enitrat replied:
Let's use Zero::zero() and One::one() (idiomatic values)

enitrat commented:
@zmalatrax I've addressed the unresolved review comments:

1. **Serde workspace dependency**: Added `serde` to workspace dependencies in `Cargo.toml` and updated `common` and `codegen` crates to use `serde.workspace = true`. Note: The prover crate doesn't currently use serde, so no changes needed there.

2. **One::one() in prover components**: Replaced all instances of `M31::from(1)` with `One::one()` in:
   - `crates/prover/src/components/multiple_constraints.rs`
   - `crates/prover/src/components/single_constraint.rs`
   - `crates/prover/src/components/single_constraint_with_relation.rs`

The other comments were already addressed:
- thiserror.workspace was fixed in commit 656a09b
- The TODO comment in generator.rs already has a reply suggesting the fix
- M31::from(0) was already replaced with Zero::zero() in the generator test

All changes have been committed in 5dd41ed.

obatirou commented on crates/common/src/instruction.rs:
```diff
@@ -0,0 +1,173 @@
+        let mut seq = serializer.serialize_seq(Some(4))?;
+
+        // Serialize as hex strings for JSON compatibility
+        seq.serialize_element(&format!("0x{:x}", self.opcode.to_u32()))?;
+        seq.serialize_element(&format!("0x{:x}", self.operands[0].0))?;
+        seq.serialize_element(&format!("0x{:x}", self.operands[1].0))?;
+        seq.serialize_element(&format!("0x{:x}", self.operands[2].0))?;
```
```suggestion
        seq.serialize_element(&format!("0x{:x}", self.opcode.to_u32()))?;
        for operand in self.operands {
            seq.serialize_element(&format!("0x{:x}", operand.0))?;
        }
```

obatirou commented on crates/common/src/instruction.rs:
```diff
@@ -0,0 +1,173 @@
+pub(crate) type InstructionArgs = [M31; 3];
+
+/// A Cairo M instruction is made of an opcode and 3 arguments.
+/// * op - The opcode id of the instruction.
+/// * args - The arguments (offsets) of the instruction.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct Instruction {
+    pub opcode: Opcode,
+    pub operands: InstructionArgs,
+}
```
```suggestion
/// * opcode - The opcode id of the instruction.
/// * operands - The arguments (offsets) of the instruction.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Instruction {
    pub opcode: Opcode,
    pub operands: InstructionArgs,
}
```

enitrat commented on crates/common/src/instruction.rs:
```diff
@@ -0,0 +1,173 @@
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct Instruction {
+    pub opcode: Opcode,
+    pub operands: InstructionArgs,
```
clippy should catch this no?

obatirou commented on crates/common/src/instruction.rs:
```diff
@@ -0,0 +1,173 @@
+}
+
+/// Deserialize instruction from JSON array
+impl<'de> Deserialize<'de> for Instruction {
```
I do not have a suggestion but it seems this could be simplified

obatirou commented on crates/compiler/codegen/src/generator.rs:
```diff
@@ -609,10 +597,10 @@ mod tests {
         assert_eq!(
             store_imm.operands,
-            vec![M31::from(42), M31::from(0), M31::from(-3)]
+            [M31::from(42), Zero::zero(), M31::from(-3)]
```
```suggestion
            [M31(42), Zero::zero(), M31::from(-3)]
```
Not sure this works for negative though, to test ?

obatirou replied:
It could be applied everywhere in test

obatirou replied:
these are tests value that are known to fit in u32 and reduce clutter but agree with your point

enitrat replied:
yeah, for example you can't do M31(-3)

obatirou replied:
that is what i though, add doubt about this one.
Anyway for others seems ok

zmalatrax replied:
You can do `-M31(3)` though, same as `M31(M31::P - 3)`

zmalatrax reviewed (approved):
Looks good to me
